### ISSUE TITLE
installer, initial page: PHP Fatal error mysqli_close(): Argument #1 ($mysql) must be of type mysqli, bool given 

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4255]
ZC158: Windows (Laragon),  mysql 5.7.32, php 8.0.3

Thrown on first page of installer (when no configures exist).
Not sure why there should be a db connection in existence at this point. Apparently there isn't one, hence the error!

> PHP Fatal error:  Uncaught TypeError: mysqli_close(): Argument #1 ($mysql) must be of type mysqli, bool given in .....\zencart-158\includes\classes\db\mysql\query_factory.php:603

https://github.com/zencart/zencart/blob/a8bef935d4d3125645ff0f727526f5cd1ae9f7f2/includes/classes/db/mysql/query_factory.php#L601-L605

 $this->link is false when the error occurs.

If  mysqli_close is commented out, the page loads.
After proceeding to second page and restoring the line, the installer proceeds as normal.

Problem does not occur with php 7.4.16.
Both php have the same mysql mode
MySQL Mode: ONLY_FULL_GROUP_BY, STRICT_TRANS_TABLES, NO_ZERO_IN_DATE, NO_ZERO_DATE, ERROR_FOR_DIVISION_BY_ZERO, NO_AUTO_CREATE_USER, NO_ENGINE_SUBSTITUTION

**Comment by drbyte** [comment]
For future reference, to make reporting issues more effective, here is feedback based on the research workflow I went through:

I can't trigger this on the "first page of installer (when no configures exist)" as described.

It's harder to investigate without a complete stack trace, particularly in this case because without it (and without being able to replicate on another PC in another environment) one can't tell what other part of the code is "calling" that function.

I was able to trigger the same fatal error at a much later stage by specifying a database name that didn't exist. In that case I get the following stack trace:
```
PHP Warning:  mysqli_connect(): (HY000/1049): Unknown database 'zencart158_torvista_test' in /includes/classes/db/mysql/query_factory.php on line 122
PHP Fatal error:  Uncaught TypeError: mysqli_close(): Argument #1 ($mysql) must be of type mysqli, bool given in /includes/classes/db/mysql/query_factory.php:603
Stack trace:
#0 /includes/classes/db/mysql/query_factory.php(603): mysqli_close(false)
#1 /includes/classes/db/mysql/query_factory.php(612): queryFactory->close()
#2 /zc_install/includes/classes/class.systemChecker.php(44): queryFactory->__destruct()
#3 /zc_install/ajaxTestDBConnection.php(26): systemChecker->runTests('database')
#4 {main}
  thrown in /includes/classes/db/mysql/query_factory.php on line 603
```
You can see here that the failing `mysqli_close()` call is not called directly by the `close()` method (like most would assume because that's where it's located, and would begin exploring there to try to figure out why it's got a wrong value) but rather by `__destruct` (which fires whenever PHP shuts down the class ... and we've been doing this since v155f), which in this case was called after entering and submitting new db credentials, via ajaxTestDBConnection.

If yours is happening on the "first page of installer", it would be interesting to know why yours is different from mine. The stack trace would help there.

Sure, the simple fix is just to skip calling mysqli_close if there's no active connection. But understanding "why" it's happening is crucial in order to be sure that there isn't another fix needed elsewhere.

**Comment by drbyte** [comment]
As you pointed out, prior to PHP 8 this wasn't triggering a fatal error, but was just (as desired) silently failing.

I'm pretty sure this is because, according to https://www.php.net/manual/en/migration80.incompatible.php, "Uncaught exceptions now go through "clean shutdown", which means that destructors will be called after an uncaught exception." and to date we're intentionally not "catching" this exception.

**Comment by drbyte** [comment]
Closed by 6f2df83a5c8008fffa764f86616b2c8902351f36

**Comment by torvista** [comment]
Stack Trace (pre-fix)
There are no configures, so no database to look for.

I'll point out that the #4 require has double slashes that don't show on quote format here ....D:\ \OneDrive\ \Dev...

> [22-Mar-2021 22:44:06 Europe/Madrid] PHP Fatal error:  Uncaught TypeError: mysqli_close(): Argument #1 ($mysql) must be of type mysqli, bool given in D:\OneDrive\Dev\public_html\zencart-158\includes\classes\db\mysql\query_factory.php:603
> Stack trace:
> #0 D:\OneDrive\Dev\public_html\zencart-158\includes\classes\db\mysql\query_factory.php(603): mysqli_close(false)
> #1 D:\OneDrive\Dev\public_html\zencart-158\includes\classes\db\mysql\query_factory.php(612): queryFactory->close()
> #2 D:\OneDrive\Dev\public_html\zencart-158\zc_install\includes\classes\class.systemChecker.php(44): queryFactory->__destruct()
> #3 D:\OneDrive\Dev\public_html\zencart-158\zc_install\includes\modules\pages\index\header_php.php(62): systemChecker->runTests()
> #4 D:\OneDrive\Dev\public_html\zencart-158\zc_install\index.php(40): require('D:\\OneDrive\\Dev...')
> #5 {main}
>   thrown in D:\OneDrive\Dev\public_html\zencart-158\includes\classes\db\mysql\query_factory.php on line 603

log 1

> Mar 22 2021 22:44 -- checkInitialSession
> PASSED
> (
>     [sessionName] => zenInstallerId
> )
> -------------
> Mar 22 2021 22:44 -- checkIniGet
> PASSED
> (
>     [inigetName] => session.use_trans_sid
>     [expectedValue] => 
> )
> ---------------
> Mar 22 2021 22:44 -- checkIniGet
> PASSED
> (
>     [inigetName] => session.auto_start
>     [expectedValue] => 
> )

log 2

> ---------------
> Mar 22 2021 22:44 -- backupConfigureFiles
> PASSED
> ---------------
> Mar 22 2021 22:44 -- CheckWriteableFile
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\includes/configure.php
>     [changePerms] => 0664
>     [createFile] => 1
> )
> ---------------
> Mar 22 2021 22:44 -- CheckWriteableAdminFile
> PASSED
> (
>     [fileDir] => includes/configure.php
>     [changePerms] => 0664
>     [createFile] => 1
> )
> ---------------
> Mar 22 2021 22:44 -- checkPhpVersion
> PASSED
> (
>     [version] => 7.3.0
>     [versionTest] => ge
>     [localErrorText] => PHP Version should be greater than 7.3
> )
> ---------------
> Mar 22 2021 22:44 -- checkPhpVersion
> PASSED
> (
>     [version] => 8.2.0
>     [versionTest] => lt
>     [localErrorText] => PHP Version should be less than 8.1.x
> )
> ---------------
> Mar 22 2021 22:44 -- checkPhpVersion
> PASSED
> (
>     [version] => 8.0.0
>     [versionTest] => ge
> )
> ---------------
> Mar 22 2021 22:44 -- checkIsZCVersionCurrent
> PASSED
> ---------------
> Mar 22 2021 22:44 -- checkExtension
> PASSED
> (
>     [extension] => mysqli
> )
> ---------------
> Mar 22 2021 22:44 -- checkExtension
> PASSED
> (
>     [extension] => pdo_mysql
> )
> ---------------
> Mar 22 2021 22:44 -- checkExtension
> PASSED
> (
>     [extension] => pdo_sqlite
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\logs
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\cache
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\images
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\includes/languages/english/html_includes
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\media
> )
> ---------------
> Mar 22 2021 22:44 -- checkWriteableDir
> PASSED
> (
>     [fileDir] => D:\OneDrive\Dev\public_html\zencart-158\pub
> )
> ---------------
> Mar 22 2021 22:44 -- checkHtaccessSupport
> PASSED
